File: model/src/main/java/net/artux/pda/model/chat/UserMessage.java
package net.artux.pda.model.chat;

import net.artux.pda.model.quest.story.StoryDataModel;
import net.artux.pda.model.user.UserModel;

import java.io.Serializable;
import java.time.Instant;
import java.util.UUID;

import lombok.Data;

@Data
public class UserMessage implements Serializable {

    private UUID id;
    private Type type;
    private UserModel author;
    private String content;
    private Instant timestamp;

    public UserMessage(UserModel userModel, String message) {
        author = userModel;
        content = message;
        timestamp = Instant.now();
    }

    public UserMessage(StoryDataModel storyDataModel, String message) {
        author = new UserModel();
        author.setLogin(storyDataModel.getLogin());
        author.setAvatar(storyDataModel.getAvatar());
        author.setPdaId((long) storyDataModel.getPdaId());
        author.setNickname(storyDataModel.getNickname());
        author.setGang(storyDataModel.getGang());

        content = message;
        timestamp = Instant.now();
    }

    public UserMessage(String senderLogin, String message, String avatarId) {
        author = new UserModel();
        author.setLogin(senderLogin);
        author.setAvatar(avatarId);
        author.setName(senderLogin);
        author.setNickname("");
        author.setRole(UserModel.Role.ADMIN);
        author.setPdaId(-1L);
        content = message;
        timestamp = Instant.now();
    }

    public static UserMessage event(ChatEvent chatEvent) {
        return new UserMessage("System", chatEvent.getContent(), "0");
    }

    public enum Type {
        OLD,
        NEW,
        UPDATE,
        DELETE
    }


}

File: model/src/main/java/net/artux/pda/model/chat/UserMessage.kt
package net.artux.pda.model.chat

import net.artux.pda.model.quest.story.StoryDataModel
import net.artux.pda.model.user.Role
import net.artux.pda.model.user.UserModel
import java.io.Serializable
import java.time.Instant
import java.util.UUID

data class UserMessage(
    var id: UUID,
    var type: Type,
    var content: String,
    var timestamp: Instant,
) : Serializable {

    lateinit var author: UserModel

    constructor(userModel: UserModel, message: String) : this(
        UUID.randomUUID(),
        Type.NEW,
        message,
        Instant.now()
    ) {
        author = userModel
    }

    constructor(storyDataModel: StoryDataModel, message: String) : this(
        UUID.randomUUID(),
        Type.NEW,
        message,
        Instant.now()
    ) {
        author = UserModel()
        author.login = storyDataModel.login
        author.avatar = storyDataModel.avatar
        author.pdaId = storyDataModel.pdaId
        author.nickname = storyDataModel.nickname
        author.gang = storyDataModel.gang
    }

    constructor(senderLogin: String?, message: String, avatarId: String?) : this(
        UUID.randomUUID(),
        Type.NEW,
        message,
        Instant.now()
    ) {
        author = UserModel()
        author.login = senderLogin
        author.avatar = avatarId
        author.name = senderLogin
        author.nickname = ""
        author.role = Role.ADMIN
        author.pdaId = (-1L).toInt()
    }

    enum class Type {
        OLD, NEW, UPDATE, DELETE
    }

    companion object {
        @JvmStatic
        fun event(chatEvent: ChatEvent): UserMessage {
            return UserMessage("System", chatEvent.content, "0")
        }
    }
}